junit-engine-api/src/main/java/org/junit/gen5/engine/support/descriptor/CompositeTestSource.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright 2015-2016 the original author or authors.
+ *
+ * All rights reserved. This program and the accompanying materials are
+ * made available under the terms of the Eclipse Public License v1.0 which
+ * accompanies this distribution and is available at
+ *
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+
+package org.junit.gen5.engine.support.descriptor;
+
+import static org.junit.gen5.commons.meta.API.Usage.Experimental;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+import org.junit.gen5.commons.meta.API;
+import org.junit.gen5.commons.util.Preconditions;
+import org.junit.gen5.commons.util.ToStringBuilder;
+import org.junit.gen5.engine.TestSource;
+
+/**
+ * A {@code CompositeTestSource} contains one or more {@link TestSource TestSources}.
+ *
+ * <p>{@code CompositeTestSource} and its {@link #getSources sources} are immutable.
+ *
+ * @since 5.0
+ */
+@API(Experimental)
+public class CompositeTestSource implements TestSource {
+
+	private static final long serialVersionUID = 1L;
+
+	private final List<TestSource> sources;
```

Comment:
I think it makes sense to preserve the order because users that are presented with the choice between multiple sources should always see the same order.

However, since we copy the input anyway (which I think is a good idea!), I think we should change the constructor to accept `Collection<? extends TestSource>`.